File: hs_src/Intake.hs
module Intake (intakeSet, intakeRun, threadIntake) where

import ConfigMap

import Vexell
import ChibiOS

import Foreign.C.Types
import Control.Monad

intakeSet :: CInt -> IO ()
intakeSet v = mapM_ (flip motorSet $ v) [leftIntake, rightIntake]

intakeRun :: IO ()
intakeRun = do
  intakeIn  <- buttonGet btnIntakeIn
  intakeOut <- buttonGet btnIntakeOut
  intakeSet $ nextState intakeIn intakeOut
  where nextState b1 b2
          | b1 = 127
          | b2 = (-127)
          | otherwise = 0

threadIntake :: IO ()
threadIntake = forever $ intakeRun >> sleep 20

File: src/intake.c
// intake.c

#include "pidbot.h"

void intakeSet(short s) {
  vexMotorSet(leftIntake, s);
  vexMotorSet(rightIntake, s);
}

void intakeRun(void) {
  if (vexControllerGet(BTN_INTAKE_IN)) {
    intakeSet(INTAKE_IN * SMAX);
  } else if (vexControllerGet(BTN_INTAKE_OUT)) {
    intakeSet(INTAKE_OUT * SMAX);
  } else {
    intakeSet(0);
  }
}

task intakeTask(void *arg) {
  (void)arg;
  vexTaskRegister("intake");

  while (!chThdShouldTerminate()) {
    intakeRun();
    vexSleep(25);
  }

  return (task)0;
}